File: cpp/algo/src/simple-algos/two-crystal-balls-problem.cpp
#include "iostream"
#include "../main.h"

using namespace std;

template <std::size_t SIZE>
int problem(const std::array<bool, SIZE> &distances) {
    const auto size = static_cast<int>(distances.size());
    const auto step = static_cast<int>(floor(sqrt(size)));

    int i = step;
    for (; i < size; i += step) {
        if (distances[i]) {
            break;
        }
    }

    i -= step;

    for (int j = 0; j <= step && i < size; j++, i++) {
        if (distances[i]) {
            return i;
        }
    }

    return -1;
}

void twoCrystalBallsProblem() {
    cout << "Two Crystal Balls Problem" << endl;

    constexpr std::array distances = { false, false, false, false, false, true, true, true, true, true };

    cout << "A bola quebra a partir de " << problem(distances) + 1 << " metros" << endl << endl;
}

File: java/kata/src/main/java/br/com/ari/algo/algorithms/TwoCrystalBallsProblem.java
package br.com.ari.algo.algorithms;

public class TwoCrystalBallsProblem {

    private TwoCrystalBallsProblem() {}

    public static int solve(boolean[] distances) {
        final int step = (int) Math.floor(Math.sqrt(distances.length));

        int i = 0;
        for (; i < distances.length; i += step) {
            if (distances[i]) {
                break;
            }
        }

        i -= step;

        for (int j = 0; j <= step && i >= 0 && i < distances.length; j++, i++) {
            if (distances[i]) {
                return i;
            }
        }

        return -1;
    }

}
